rfd/0014-custom-approval-conditions.md

```diff
@@ -0,0 +1,378 @@
+doesn't stop us from supporting a loose concept of *suggested* reviewers.
+
+When a user generates an access request, we can accept a list of arbitrary strings identiying
+suggested reviewers:
+
+```
+$ tsh login --request-roles=dictator --request-reviewers=alice@example.com
+```
+
+Rather than requiring `alice` and `bob` to be existing teleport users, we can simply store
+the strings alongside the request.  This will allow `tctl` and the Web UI to filter requests
+by arbitrary reviewer strings (including the name of the currently logged in user if
+applicable).  Ex:
+
+```
+$ tctl request ls --wants-review-from=alice@example.com
+# ...
+```
+
+Simultaneously, plugin implementations may use this field in other ways.  A slack plugin,
+for example, could DM each slack user whose name appeared in the suggested reviewers
+list, providing a link to the request within the Web UI.  Since the request ID is opaque,
+we don't actually need to verify any permissions until the link is clicked and the
+potential reviewer authenticates with teleport.
+
+If manually listing reviewers ends up being tedious, it would also be feasible to hard-code
+suggested reviewers within an `allow.request` block like so:
+
+```yaml
+kind: role
+# ...
+spec:
+  allow:
+    request:
+      suggested_reviewers: ["alice@example.com", "bob@example.com"]
+      # ...
+```
+
+#### Supporting Extended Options
+
+Ideally, all options available in the old approval style should be available in thresholded
+approvals (i.e. an approval threshold of `1` should not be a special case).  Unfortunately,
+not all parameters supported by the existing system translate intuitively to a multi-party
+system:
+
+
+##### Annotations
+
+Calculating the `ResolveAnnotations` (annotations supplied upon approval/denial) of a request
+is a bit tricky.  If two reviews provide two different annotation mappings, do you pick one?
+Simply using the annotations from either the first or last review is problematic because that
+makes us very sensistive to ordering (potentially leading to very confusing bugs).
+
+Given that annotations are of the form `map[string][]string`, the annotations of all
+reviews could theoretically be "summed" (e.g. `{"hello": ["world"]}` and `{"hello": ["there"]}`
+become `{"hello": ["there","world"]}`).  This isn't a perfect solution, as it would prevent
+users from treating the order of annotations as meaningful, but that may be for the best.
+They were never intended to be meaninfully ordered, and this same kind of summing already
+happens elsewhere.
+
+
+##### Role Overrides
+
+Role overrides are very tricky.  The existing system allows approvers to override the list of
+roles granted by an access request (specifically, approvers can subselect, they cannot add
+roles which are not present).  How this should map to thresholded approvals is unclear.
+
+Say that `dave` requests roles `["foo","bar","bin"]` with an approval threshold of `2`.
```

Comment:
If `dave`'s approval requests were to be replicated for a single role giving 3 requests for `foo`, `bar` and `bin` respectively with the corresponding threshold (`2`), it could be easier to reason about. In this case, `bob` and `alice` would both be able to approve `foo` and `bar` but `bin` but would be left unapproved, so `dave` would have to wait for someone else with a permission to approve `bin` to successfully gain access.

Reply:
This would make the processing much more complex though - right now the request is a transaction with a state, it's either approved or not. With your proposal, the request would be split into multiple object, each with it's own state. This would be quite hard to reason about, as request could be partially approved/denied.